### PR TITLE
fix(@clayui/color-picker): Current colour persist and Hue component synchronization

### PR DESCRIPTION
**Pull request by sergiojimcos**
**What is new?**
There are some changes:
-  It saves the current colour in the **Custom Colour Palette**. 
- **Hue** synchronization with the current color and the rest of components.
- **Input** elements in upper case.

## WITH PALETTE

https://user-images.githubusercontent.com/44723449/155372894-1d37400b-86ed-440e-bcf5-bd3a0a3fbfc0.mov

## WITHOUT PALETTE

https://user-images.githubusercontent.com/44723449/155372964-3ad1e707-b7c5-429e-8696-24bc3dfa8abd.mov

**FIXES**
#4580

### CODE REVIEW

File: packages/clay-color-picker/src/Custom.tsx
Comment (matuzalemsteles) [inline packages/clay-color-picker/src/Custom.tsx:193]: It seems that this is having a side effect when interacting with the Hue component and the Gradient selector when moving to a black color and in some situations the Hue is also changing color when moving through the Gradient.

See this video:

https://user-images.githubusercontent.com/13750819/155434784-ceb13a73-785e-4a16-a078-e1457e990f1f.mp4

**Review by matuzalemsteles — COMMENTED** [review]
Hey @sergiojimcos I think we are deviating from the expected behavior here, apparently when color is added to the input that is not a color that exists when opening the panel the color should only be reflected in the editor but not modify the custom color but if the user clicking on an empty color in the custom color will replace it with the color added by the user.

So if color is manually added in the input that doesn't exist in the custom color, when opening the panel it should not focus/active on a custom color.

See this comment by Emiliano https://github.com/liferay/clay/issues/4580#issuecomment-1034734204. Let me know if you need help!

**Review by matuzalemsteles — COMMENTED** [review]
I've noticed some inconsistent behavior for some scenarios, but I think we're getting there.

- When typing on the input the Panel is being closed it must remain open
- The input value I think we don't need to leave uppercase for all scenarios, for example `var(--blue)` this can break the implementation, also it's ok to type the color name in lowercase when typing a lowercase hex color is also not a problem.
- Typing a color initially in the input and then opening the panel and clicking on an empty color is not replacing the color typed in the input

File: packages/clay-color-picker/src/__tests__/__snapshots__/index.tsx.snap
Comment (julien) [inline packages/clay-color-picker/src/__tests__/__snapshots__/index.tsx.snap:1236]: @sergiojimcos do you know why this needs to be updated? not saying it's wrong but I'm curious about the change.

File: packages/clay-color-picker/src/index.tsx
Comment (julien) [inline packages/clay-color-picker/src/index.tsx:348]: @sergiojimcos looks like you forgot to "clean up" here

File: packages/clay-color-picker/src/Custom.tsx
Comment (julien) [inline packages/clay-color-picker/src/Custom.tsx:260]: I'd rather rename `i` to `index` in this scope just to make it clearer

**Review by julien — COMMENTED** [review]
Looks OK overall, I'm just going to fetch this to see it in an editor, GitHub's UI isn't the greatest thing to review

**Comment by julien** [comment]
@sergiojimcos before submitting a pull request or updating it, you can always run the tests locally to make sure they won't fail

File: packages/clay-color-picker/src/__tests__/__snapshots__/index.tsx.snap
Comment (sergiojimcos) [reply packages/clay-color-picker/src/__tests__/__snapshots__/index.tsx.snap:1236]: Sure @julien. At first I changed this snapshot when I modified the values of the input into upper case. As I told you yesterday, finally the value of the input must be in lower case, for that reason I changed into into the default one

File: packages/clay-color-picker/src/index.tsx
Comment (sergiojimcos) [reply packages/clay-color-picker/src/index.tsx:348]: Oh sorry, I haven't seen that 😅

**Comment by julien** [comment]
@sergiojimcos I've updated everything we did together during our pairing session in https://github.com/liferay/clay/pull/4691/commits/505253aaa0226765be8e988433fe8a77751784c9, let's see if we're luck with the tests :wink:

**Comment by sergiojimcos** [comment]
> @sergiojimcos I've updated everything we did together during our pairing session in [505253a](https://github.com/liferay/clay/commit/505253aaa0226765be8e988433fe8a77751784c9), let's see if we're luck with the tests 😉

Great, many thanks Julien! ❤️

**Comment by julien** [comment]
@matuzalemsteles I helped @sergiojimcos a bit with this and something isn't very clear:

Do we want to let users be able to use things like CSS variables (e.g. `'var(--blue)'`) and "invalid" colors, like `'inherit'`?

If that's the case then we probably need to think about this very carefully, because `tinycolor` doesn't consider those to be "valid" colors, (i.e. `tinycolor('inherit').isValid()` will return false, same with `var(--blue)`). Additionally we also have this [hook](https://github.com/liferay/clay/blob/5591614c1b88a799eab9cc3afbd96c986180faae/packages/clay-color-picker/src/hooks.ts#L38-L56), that not going to work with those values. 

What are your thoughts on this?

**Comment by matuzalemsteles** [comment]
Yeah, it's intentional to support this, this came up with this issue #3681 as well to help support use in stylebooks. We'll probably have to improve our implementation of this to identify when it's using different syntaxes.

Probably for CSS Variables cases, we will have to get the hex just for internal use to show in the gradient and the hue. Apparently, we have the same with the `inherit` and named colors using something like `getComputedStyle(ref).backgroundColor` which returns the rgb, maybe we can try something with that, I would say don't do anything with this here, maybe just avoid uppercase for this scenario and create another issue to work on, we will probably have to rewrite some things I think.

Maybe we can accept any value typed in the input for now so we don't have to use `tinycolor('inherit').isValid()` for example.

**Comment by julien** [comment]
@matuzalemsteles I know this is not "your decision" but  this "feature" doesn't really make any sense ... 

I also think that since @sergiojimcos is new on the team we should try to help him as much as we can when it comes to task definitions and goals - so I'll try to help him on this one.

**Comment by matuzalemsteles** [comment]
> Do you know if this request come from Lexicon or Dialect?

I'm not sure but tracing the conversations, this stemmed from the need for stylebooks.

> It's possible to get the color value when the color is set to inherit, but we'll need to modify the component to add an "alpha input", because for the moment we only support RGB (and not RGBA)

We have a request from Dialect to support RGBA #4134 but we are depending on the Lexicon team because it will need changes to the Panel to support this.

Maybe what we can do here is convert rgba to rgb, I think the `tinycolor` lib can help us with that. It also looks like it can't always come rgba from `inherit` it will depend on the parent declaration.

> We need to get rid of the [useHexInput](https://github.com/liferay/clay/blob/c62913222b959af3ce13f0e00cd10a8406ab4a33/packages/clay-color-picker/src/hooks.ts#L38-L56) hook because it's too restrictive for our this case

Hmm, the problem is that the panel's input works with hex, maybe if it reflects the same thing that is in the input it can make sense to remove it.

> Last but not least, since also need to be able to work with CSS variables, we probably want to add a few CSS variables to the stories in order to check if we can use something like [this](https://developer.mozilla.org/en-US/docs/Web/CSS/Using_CSS_custom_properties#values_in_javascript) to retrieve the correct values

Yeah, I think there's a case of `var(--blue)` in the custom color storybook example.

**Comment by julien** [comment]
> I'm not sure but tracing the conversations, this stemmed from the need for stylebooks.

Yes, that's also what I imagined. 

> Maybe what we can do here is convert rgba to rgb, I think the tinycolor lib can help us with that. It also looks like it can't always come rgba from inherit it will depend on the parent declaration.

Yes, we can try that although I'm not convinced it will work with all use cases. 
I agree that sometimes `inherit` won't always be an RGBA value, but sometimes it will and it also seems that we want to support "anything"

> Hmm, the problem is that the panel's input works with hex, maybe if it reflects the same thing that is in the input it can make sense to remove it.

I understand that we want the "main input" to accept either RGB, RGBA, CSS variables (and eventually color "names", like red, blue, green, etc...) and that the input in the panel should show those values converted to hex; at the same time we also want to support RGBA at one point, and that hook won't let us do that. Additionally since we already have `tinycolor` that can validate the color, having a hook based on regular expressions doesn't seem necessary ... We should have one way of validating the color, not two.

**Comment by julien** [comment]
@matuzalemsteles as far as I can see, there's no way to convert from RGBA to RGB with `tinycolor` - and since we're still waiting for the design changes I guess it's better to put this on hold (although the bug was fixed)

**Comment by matuzalemsteles** [comment]
> Additionally since we already have tinycolor that can validate the color, having a hook based on regular expressions doesn't seem necessary ... We should have one way of validating the color, not two.

Yeah, that makes sense to me too, I think this was created just for a way of hex validation, but could have used `tinycolor` for that purpose. Well, I think we can remove that and just use `tinycolor`.

> as far as I can see, there's no way to convert from RGBA to RGB with tinycolor - and since we're still waiting for the design changes I guess it's better to put this on hold (although the bug was fixed)

yeah, it makes sense he will probably put an 8 digit hex that we don't support yet 😔.

**Comment by sergiojimcos** [comment]
Hi @matuzalemsteles! As @julien explains before, I have removed `useHexInput` hook and use `tinycolor` instead to avoid comparing using the regular expression. According to 8 digit hex, I think that we can see later after resolving this issue and investigate how can we support RGBA in color-picker once it works well. Yo can review this.

**Comment by matuzalemsteles** [comment]
Hey @sergiojimcos I'm reviewing this, tomorrow I'll add some interactions on top of this PR to cover some edge cases, no need to worry.

**Comment by matuzalemsteles** [comment]
Hey guys, I pushed some modifications in an attempt to simplify some scenarios and also cover the use cases that was requested by Emiliano. So in these changes we have the following changes:

- Typing a different color in the input than the one in the custom palette and opening the Panel will not overwrite any color until you click on an empty slot
- Fix the problem of typing in the custom input was not modifying the color
- When opening the panel with a color different from the custom color, it is reflected in the editor and it is also possible to change the color without having to save it in the palette
- Adds support for CSS Variables, color is converted to 6-digit Hex in editor so user can view or change

About the case of `inherit` I was thinking about it and it doesn't make much sense we support it in the editor and the slot too, despite the possibility to get the color of `inherit` it doesn't make sense because the `inherit` it will be used in another place that we have no way to infer what the color of the parent will be where it is being used, for example, I think that in this case, the product must have a different behavior for that.

**Comment by julien** [comment]
Hey @matuzalemsteles I think these changes are *almost* OK, I do agree that supporting `inherit` doesn't make much sense as it can be complex to handle - to clarify, I used *almost* because everything work except for when clicking on the white (`FFFFFF`) squares - for some reason the "main" input doesn't get updated. Appart from that LGTM

**Comment by matuzalemsteles** [comment]
Hey @julien could you say what exactly this behavior would be? are you meaning this behavior? Apparently, it's working as expected, I may be missing some other use case.

https://user-images.githubusercontent.com/13750819/157959367-ee98576d-b698-4ecc-b9a6-4b3aa40205a1.mp4

**Comment by sergiojimcos** [comment]
Hey @matuzalemsteles! What @julien means in this scenario is that when we click on a `FFFFFF` splotch, both inputs must be updated. Apart from that, everything is great, LGTM! 

https://user-images.githubusercontent.com/44723449/158130293-baaad763-222e-4e71-b0f7-cfac5033cf83.mov

**Comment by matuzalemsteles** [comment]
Hey guys, I've fixed this use case, let me know if I'm missing anything else. I will also create an issue so we can add more test cases for ColorPicker.

**Comment by sergiojimcos** [comment]
Hey @matuzalemsteles! I've updated the branch and this (https://github.com/liferay/clay/pull/4691#issuecomment-1066486933) is solved and works well.

However, I've been tested several use cases to check that color picker finally works well and it doesn't. There are two use cases that are related:

1. In this case, I think that color picker must store only one value of a particular color, because it doesn't make sense having several instance of the same color. I don't remember well but this particular example was discussed. 

https://user-images.githubusercontent.com/44723449/158224317-702e4796-1dc5-47ce-941e-a95a616564ab.mov

2. On the other hand, it happens only with CSS variable. I think that in this case, the color picker doesn't check the hex values of the CSS variables, because the second value is stored as hex value

https://user-images.githubusercontent.com/44723449/158225248-09f68863-42e3-4ad1-a7ef-8712dd42f83b.mov

**Comment by matuzalemsteles** [comment]
hey @sergiojimcos Thanks for the investigations, I'll work on it! It makes sense to constrain the splotch to not have repeating colors.

**Comment by matuzalemsteles** [comment]
I pushed more changes in this PR to fix the above scenarios, I did several tests too to check if it's working well, but I believe I may have missed something, I also tried to simplify the implementation more and add some comments to help in understanding the edge cases.

I believe that a lot of these problems are a little related to the way the components are structured, Custom is depending a lot on the `value` to define the values, so it's more of a state management problem that is hurting a lot here to define the other states. At another time we can simplify and refactor this component to have better management of the internal states to simplify the final implementation.

**Review by julien — APPROVED** [review]

**Comment by julien** [comment]
@matuzalemsteles LGTM thanks. @sergiojimcos could you have a second look at this before we merge?

**Comment by sergiojimcos** [comment]
LGTM! thanks @matuzalemsteles!

**Pull request merged**
